rfc-115-enabling-http2-on-govuk.md

```diff
@@ -26,67 +29,95 @@ The RFC below discusses the problems with our current setup and suggests possibl
 
 > `Access-Control-Expose-Headers`, `Access-Control-Allow-Methods`, and `Access-Control-Allow-Headers` response headers can only use `*` as value when request’s credentials mode is not "include".
 
-The use of the wildcard (`*`) isn't allowed on a credentialed connection. If used it will block any requests and post an error message to the browser that looks like this:
+Basically the use of the wildcard (`*`) isn't allowed on a credentialed connection (`use-credentials`). And if it is used the browser will block any requests and post an error message to the that looks like this:
 
 > Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at ‘https://assets.example.com/script.js’. (Reason: Credential is not supported if the CORS header ‘Access-Control-Allow-Origin’ is ‘*’).
 
-This is caused by our current NGINX setup on [these lines](https://github.com/alphagov/govuk-puppet/blob/962ea899e9c6778fe91e80074346912bd4314b10/modules/router/templates/assets_origin.conf.erb#L36-L38).
+We are currently [serving all our assets](https://github.com/alphagov/govuk-puppet/blob/962ea899e9c6778fe91e80074346912bd4314b10/modules/router/templates/assets_origin.conf.erb#L36-L38) from the `assets` domain with the `Access-Control-Allow-Origin: *` header, which is blocking us from making this change. The reason why we are serving assets like this is because these assets are actually being used on other domains. 
+
+The Heroku applications (e.g. https://government-frontend.herokuapp.com) that are created are a legitimate reason for using this header. If we were to remove the `Access-Control-Allow-Origin: *` header then the JS and fonts (due to their [unique CORS requirements](https://www.w3.org/TR/css-fonts-3/#font-fetching-requirements)) on the Heroku previews would break. According to the `Access-Control-Allow-Origin` [documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin), there are only 3 valid values for this header:
+
+- `*`
+- `<origin>`
+- `null`
 
-We should look at assets served using this header. Note: Webfonts will need to be served with this header due to their [unique CORS requirements](https://www.w3.org/TR/css-fonts-3/#font-fetching-requirements).
+The `<origin>` value is very important. As this is where we would ideally like to specify: `*.gov.uk; *.herokuapp.com`. Unfortunately this header only accepts a **single** origin, and it **doesn't** recognise wildcard values. So we are in a situation where we need to use `*` (for SRI) and a specific origin value (e.g `https://government-frontend.herokuapp.com` for every unique Heroku app URL) at the same time. As far as I know this isn't possible to fulfil, so the `Access-Control-Allow-Origin: *` header must stay as is.
```

Comment:
There's a Fastly blog post that suggests it might be possible to dynamically set the `Access-Control-Allow-Origin` header based on `req.http.Origin` at the CDN level – https://www.fastly.com/blog/caching-cors

Reply:
Excellent thanks @36degrees 👍 

Depending on the complexity involved and the number of external domains the assets are used on, this should allow us to use SRI with `crossorigin='use-credentials'`.

Reply:
It'd be a bit tricky as we'd have to set-up dynamic hostname responses - as the heroku apps are generated based on pull requests. We'd also need similar configuration for users developing locally so we'd need the same again for dev.gov.uk hosts. I'm not sure if we need to set-up anything for localhost with a port though, does CORS treat localhost special?

I'd suggest we'd want to be really clear that we're benefiting from SRI before complicating the configuration to support it. It sounds like this RFC has determined we're not using it for the purpose it's intended - have we received any push back on this suggestion?

Reply:
No pushback as of yet. Personally I'm in favor of keeping the `Access-Control-Allow-Origin: *` on our assets to allow Heroku (and any other services using the assets to keep working), and remove SRI. As you say it isn't being used as intended (third-party scripts) and the path of least resistance / complications is removing it.

Reply:
> I'd suggest we'd want to be really clear that we're benefiting from SRI before complicating the configuration to support it.

👍 

> It'd be a bit tricky as we'd have to set-up dynamic hostname responses - as the heroku apps are generated based on pull requests.

I'm not terribly familiar with VCL, but could you match against a regex to mimic wildcard support for e.g. `*.herokuapp.com`? Something like…

```
if (req.http.Origin ~ "\.herokuapp.com$") {
    set resp.http.Access-Control-Allow-Origin = req.http.Origin;
  }
```